### PR TITLE
Add an ADR for our "threadsafe interfaces" plan.

### PR DESCRIPTION
**Pull request by rfk**
### [Click for Rendered View](https://github.com/mozilla/uniffi-rs/blob/adr-threadsafe-interface/docs/adr/0003-threadsafe-interfaces.md)

Based on conversations over the last few days, here is a concrete proposal for how we can resolve the underlying issue that casued the [recent Fenix startup regression](https://github.com/mozilla/uniffi-rs/blob/adr-threadsafe-interface/docs/adr/0003-threadsafe-interfaces.md). I hope I have captured the current consensus accurately, but this is an ADR, so please do weigh in if you don't agree with something here or if you've got other considerations to include!

Something new for this ADR: I have included an [implementation sketch](https://github.com/mozilla/uniffi-rs/blob/adr-threadsafe-interface/docs/adr/0003-threadsafe-interfaces.md#implementation-sketch) section with a high-level overview of how I'd go about implementing this. I'm hoping this will empower someone who is not me to actually implement this and learn a bit more about this part of the uniffi codebase.

### CODE REVIEW

**Comment by rfk** [comment]
Optimistically assuming approval for the plan proposed here, I've also filed https://github.com/mozilla/uniffi-rs/issues/370 to track the work of implementing it.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (jhugman) [inline docs/adr/0003-threadsafe-interfaces.md:180]: I expect we're going to have to delve into the method bodies themselves, to add or remove `mut` annotations in various places.

**Review by jhugman — APPROVED** [review]
This looks like a really nice path forward.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (rfk) [reply docs/adr/0003-threadsafe-interfaces.md:180]: Yes, I think you're right; we may even find we want to define a small `HandleMap` trait of our own and adapt `ConcurrentHandleMap` to it, rather than tying ourselves to to the API it already offers.

**Review by dmose — COMMENTED** [review]
Any chance of word-wrapping this more narrowly?  As is, making it readable in the comment view of a github diff window means dealing with very tiny fonts.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (dmose) [inline docs/adr/0003-threadsafe-interfaces.md:164]: How about 'LessLockingHandleMap' or 'MinimalLockingHandleMap', since there _is_ still a lock in there, and projects that don't require a singleton (i.e. not Nimbus) could still potentially block on locks during object construction/destruction?  If this is too edge-casey to scare all code readers about, it'd be nice to at least document the possibility in the reference.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (dmose) [inline docs/adr/0003-threadsafe-interfaces.md:187]: I would assert that we don't really want this to be optional, in the sense that I think there's high value in (at the very least) encouraging method implementors to document if they're going to make something blocking as part of their contract with the API consumer.

That said, I almost want to suggest different semantics: have a developer affordance to require the developer to explicitly think through and document what their intent is at the time the code is written.  Which is to say, have [Blocking] and [NonBlocking] attributes, and disallow attribute-free methods.

**Review by dmose — COMMENTED** [review]
Note that I'm still new to Rust (i.e. I just learned about `Sync` and `Send` in order to do this review), so there's a decent bit of stuff here that I don't have tons of context around (eg `Box` proposals).  So definitely don't land this with just my review.  :-)

**Comment by rfk** [comment]
> Any chance of word-wrapping this more narrowly?

I'll give it a shot, but also, that's why there's a link to the rendered view in the issue comment 😁

File: docs/adr/0003-threadsafe-interfaces.md
Comment (rfk) [reply docs/adr/0003-threadsafe-interfaces.md:187]: Oh, interesting suggestion. I think I'd be OK with trying that out as a required annotation and see if we like it, as we can always remove the restriction later.

Comment (dmose) [reply docs/adr/0003-threadsafe-interfaces.md:187]: Indeed; that's the nice thing about only having one real client -- it's still pretty easy to experiment with stuff like this.  :-)

Comment (dmose) [reply docs/adr/0003-threadsafe-interfaces.md:187]: I guess constructors would need to be declared as [Blocking], since there is gonna be some locking around the handlemap...

Comment (rfk) [reply docs/adr/0003-threadsafe-interfaces.md:187]: I'm not sure what we intend the semantics of `[Blocking]` to mean, but subjectively, there feels like a significant difference between "blocks on network or disk I/O" and "takes a mutex protecting a purely-in-memory data structure". *Any* method call on an object is going to have to interact with the handlemap mutex, so I worry that considering handlemap access a blocking operation will eliminate any documentary value that the annotations may provide.

Comment (rfk) [reply docs/adr/0003-threadsafe-interfaces.md:187]: By contrast, I think there *is* value in being able to know whether or not the constructor does disk I/O via some sort of annotation.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (rfk) [reply docs/adr/0003-threadsafe-interfaces.md:164]: FWIW I will not block (hah!) this on the selection of any particular name; all of the suggestions so far are acceptable to me and I'll be happy to have a verbose name if it conveys an accurate mental model of how the thing behaves. (Which I feel like `ConcurrentHandleMap` doesn't necessarily do - is it "concurrent" because it's safe to use with concurrent access, or because it enables lots of concurrency?)

File: docs/adr/0003-threadsafe-interfaces.md
Comment (mhammond) [inline docs/adr/0003-threadsafe-interfaces.md:67]: Not mentioned here is that it's a *requirement* for nimbus which would have been impossible to meet if we did nothing.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (mhammond) [inline docs/adr/0003-threadsafe-interfaces.md:32]: I had a mild reaction to "non-locking handlemap" as it might imply no locks are necessary at all, but I expect locks will be necessary to protect the handlemap itself? Not sure it's worth trying to squeeze this in though.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (mhammond) [inline docs/adr/0003-threadsafe-interfaces.md:90]: "Good, because it leaves the door open to "Option C" in the future if we decide that makes sense later."?

File: docs/adr/0003-threadsafe-interfaces.md
Comment (mhammond) [reply docs/adr/0003-threadsafe-interfaces.md:164]: Agreed, the existing name is itself bad - so my 2c would be something like "NonConcurrent..." - ie, if you understand how the term "concurrent" applies to the existing one, you automagically understand what's different about the new one.

File: docs/adr/0003-threadsafe-interfaces.md
Comment (mhammond) [reply docs/adr/0003-threadsafe-interfaces.md:187]: I agree with all the above, but in the short term, I believe the main value is documenting expectations, so would like to see *something* here sooner rather than later (eg, we will have `getExperimentBranch()` which doesn't "block", but now there's a `getExperimentBranches()` function which, IIUC, does *not* need these semantics - which will not be at all obvious without some url annotation

**Review by mhammond — APPROVED** [review]
Lovely! Most of my comments don't require changes, but I thought them worth noting.

**Comment by mhammond** [comment]
A random thought that will probably go nowhere:

For the sake of the strawman, let's say the annotation on the method could be [Blocking] and [Not-Blocking]. Any function marked as [Blocking] is expected to take a lock. It's not clear how uniffi might be able to help here - taking a mutex on those functions isn't that helpful as the entire object needs to be Send+Sync, and this isn't really the right granularity. But maybe there could be another Rust method called, and *that* is passed in.

To make this clearer - in the nimbus PR, we have something like:

```rust
struct Nimbus {
  db: Mutex<Database>
}

impl Nimbus {
  pub fn blocking(&self, ...) {
    let db = self.some_mutex.lock().unwrap();
    ...
  }
  pub fn not_blocking(&self, ...) {
    ... // doesn't take the mutex
  }
}

```

Whereas I can imagine something like:
```rust

struct ProtectedData {
  db: Database,
}

struct Nimbus {
  data: Mutex<ProtectedData>
}

impl Nimbus {
  // called by uniffi for all functions marked as [Locking] and the result is passed as the 2nd param to the functions
  fn uniffi_lock(&self) -> Handwave {
    self.data.lock.handwave()
  }

  pub fn blocking(&self, data: &ProtectedData, ...) {
  }
  pub fn non_blocking(&self) {
  }
}
```

and uniffi still manages the mutex for us. Or something.

1/2 baked, and we certainly don't want to block (ha!) on this, but there might be a grain of a way nimbus can help enforce semantics and reduce boilerplate. A key limitation is that it removes flexibility, so we might *still* end up needing a way to overwrite the semantics if someone truly has free-threading requirements, or at least invent a way that the app itself is in charge of the semantics.

**Comment by dmose** [comment]
A very interesting idea; would love to hear others' thoughts on it...

File: docs/adr/0003-threadsafe-interfaces.md
Comment (rfk) [inline docs/adr/0003-threadsafe-interfaces.md:161]: Ref https://github.com/mozilla/uniffi-rs/pull/372#discussion_r558620397, this type is not sufficient because I don't think Rust will let us release the `RwLock` while there is an outstanding reference to an item from the map. As well it shouldn't: if we let one thread operate on an `&T` while a concurrent thread deletes the underlying `T` from the handlemap, we'd risk running `T::drop` while the `&T` was still alive.

I expect this will need to be `RwLock<HandleMap<Arc<T>>>` so that we can reference-count our way around this issue. Leaving a note for myself to update the final ADR before merge.

**Comment by dmose** [comment]
I've filed a separate issue (#378) on the [Blocking]/[Non-Blocking] implementation itself.  Maybe the specific proposal wants to be nailed down before this PR is landed, or maybe not.

**Comment by rfk** [comment]
I've updated the doc to fold in the comments above, and added links to the implementation PR and follow-up issue from `[Blocking]`. I'll go ahead and merge this once CI gives me the all-clear. Thanks folks!

**Pull request merged**
